server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java

```diff
@@ -112,8 +116,11 @@ protected String contentType() {
     public void postParse(DocumentParserContext context) {
         // Ignored values are only expected in synthetic mode.
         assert context.getIgnoredFieldValues().isEmpty() || context.mappingLookup().isSourceSynthetic();
-        for (NameValue nameValue : context.getIgnoredFieldValues()) {
-            context.doc().add(new StoredField(NAME, encode(nameValue)));
+        List<NameValue> ignoredFieldValues = new ArrayList<>(context.getIgnoredFieldValues());
+        // ensure consistent ordering when retrieving synthetic source
+        Collections.sort(ignoredFieldValues, Comparator.comparing(NameValue::name));
```

Comment:
Is this more efficient than using a `TreeSet`?

Reply:
With the new approach we may have collisions so I had to switch to a `List` instead. 
Consider for instance:
```
{
    "nested": [
        {
            "obj": {
                "field": "bar"
            }
        },
        {
             "obj": {
                "field": "bar"
            }
        }
    ]
}
```
where `nested` is a nested field and `obj` is a disabled object field (with `enabled=false`) within.
In such case we'll register the `obj` field twice as an ignored field, one for each entry in the nested array but targeting different Lucene document.

Reply:
Interesting.. Makes sense, would we also need to sort by doc id after name?

Reply:
I don't think this is needed since fields with the same name should be on different nested documents.